### DISCUSSION TITLE
OpcUa Server how to get Client write event instead of change data event

### DISCUSSION THREAD

**Discussion by Dorammon** [FreeOpcUa/python-opcua#1398] [Q&A]
###
count = 0
where True:
  count += 1
  time.sleep(1)
###
"count" is OpcUa server attribute, delay 1 second plus 1, I want get client write event in server instead of  the data change event
i don't know how to do , i see the example code not found

**Comment by AndreasHeine** [comment]
you probably need to explain a little more what you want to do!
you want a callback if a value changes?

**Reply by Dorammon** [reply]
Hi
if value changes need callback, don't change the value every time callback, only client write value need callback

**Reply by Dorammon** [reply]
I need the client write callback, to do something

**Reply by AndreasHeine** [reply]
sadly there is no client write callback implemented!

Following the incomming request:
https://github.com/FreeOpcUa/python-opcua/blob/3012acf9047aad3acc2674e9008f0d1eee680c25/opcua/server/uaprocessor.py#L197
https://github.com/FreeOpcUa/python-opcua/blob/3012acf9047aad3acc2674e9008f0d1eee680c25/opcua/server/internal_server.py#L362
https://github.com/FreeOpcUa/python-opcua/blob/3012acf9047aad3acc2674e9008f0d1eee680c25/opcua/server/address_space.py#L53
https://github.com/FreeOpcUa/python-opcua/blob/3012acf9047aad3acc2674e9008f0d1eee680c25/opcua/server/address_space.py#L662

**Reply by Dorammon** [reply]
OK, thank you!

**Reply by AndreasHeine** [reply]
you could probably get away with storing the last value written from the server and compare it with the uavariable value if it is different its changed by the client (but the comparing needs to be inside a valuechange callback)

**Reply by Dorammon** [reply]
Sorry for your late reply
![企业微信截图_16339139023984](https://user-images.githubusercontent.com/3895690/136719609-51d962e9-99d6-4f60-aac8-76b761456550.png)

![企业微信截图_16339143859472](https://user-images.githubusercontent.com/3895690/136719928-bfe78a35-5b6c-45d6-8d27-00c563ae91a2.png)

Is that what it means？
Sometimes this callback will not be detected


Can I judge in data change callback that it is written by the client?

**Reply by AndreasHeine** [reply]
no! i am talking about a valuecallback in the server not on clientside.
